### ISSUE TITLE
Streamline IDE Integration - cbuild setup command -

### ISSUE THREAD

**Issue by ReinhardKeil** [Open-CMSIS-Pack/devtools#1265]
**Describe The Problem To Be Solved**

Any change of csolution projects in VS Code requires:
* Validation of the modification; does it work in with all context settings
* Creation of `compile_commands.json` for IntelliSense features; this could be limited to the current active context. This is currently generated by a cbuild -t database (#1210) using a special CMake target.

The `compile_commands.json` should be also generated at startup of VS Code to ensure that it is up-to-date.

Currently there are two commands issued on start of VS Code to validate (csolution convert) and build `compile_commands.json` (cbuild). Both do effectively the same things.

In addition there is a problem with `compile_commands.json` when the out-dir of a multi-project application is configured (as it is the case with https://github.com/Open-CMSIS-Pack/csolution-examples/tree/main/DualCore).

**Suggest A Solution**
- Run "cbuild setup" as explained here: https://github.com/Open-CMSIS-Pack/devtools/issues/1265#issuecomment-1887396374
- Solution Outline View in an IDE could be generated as below:
  - Read csolution.yml – gives information about all projects/target/build types (used for build control)
  - For a selected context – read the cproject.yml file – show group/user source files in outline view
  - In case that the cproject.yml contains layers, read clayer.yml – show group/user source files.
  - List all Cclass names that are in cproject.yml/clayer.yml - use this also to capture where the component is defined.
  - For getting the Cclass source files, open the build information file of the [current context](https://github.com/Open-CMSIS-Pack/cmsis-toolbox/blob/main/docs/YML-CBuild-Format.md#file-structure-of-cbuild-yml)
     - This file contains all source files that belong to components.
     - Further information is in [23wk48_Open-CMSIS-Pack_TM.pdf](https://github.com/Open-CMSIS-Pack/devtools/files/13963649/23wk48_Open-CMSIS-Pack_TM.pdf)
- Add Generator run information:
  - For components that require a generator run, the build information file is extended with:
     ```
     components:
        - component: Keil::Device:CubeMX@0.9.0
          run: CubeMX
          gen: <file>.cgen.yml
     ```
  - This information is used by the IDE to:
     - show a "Run CubeMX" button in the Solution Outline View
     - watch for changes to `<file>.cgen.yml` which retriggers a csolution setup
  
- Other information about API Headers, User Code Templates, Documentation could be added in a similar way.

**Comment by mcgordonite** [comment]
> In addition there is a problem with compile_commands.json when the out-dir of a multi-project application is configured (as it is the case with https://github.com/Open-CMSIS-Pack/csolution-examples/tree/main/DualCore).

https://github.com/Open-CMSIS-Pack/devtools/issues/1210 would be a good fix for this: the VS Code extensions could then read the compile_commands.json location from the cbuild.yml file.

**Comment by brondani** [comment]
> #1210 would be a good fix for this: the VS Code extensions could then read the compile_commands.json location from the cbuild.yml file.

This information can already be implied from cbuild.yml by appending `compile_commands.json` to the `outdir`.
For example:
```
output-dirs:
    outdir: out/hello/B-U585I-IOT02A/Debug
```
The compilation database is generated in `out/hello/B-U585I-IOT02A/Debug/compile_commands.json`

**Comment by ReinhardKeil** [comment]
What about extending the `cbuild` command with support for IDE.  In IDE context the assumption is that we always use a context-set, but perhaps we can make this less strict.

```
cbuild setup <name>.csolution.yml --IDE { VSC_clang | VSC_msft } -S
```
Command `setup` would verify that the YML files are correct. Internally it uses potentially the `csolution convert` command to generate build information files (even in case of errors).  This command is called at the start of an IDE or whenever a related YML input file of the selected csolution project is changed.

Option `--IDE` generates IDE setup files. For VS Code for each context a `compile_commands.json` is generated in the related [`tmp` directory](https://github.com/Open-CMSIS-Pack/cmsis-toolbox/blob/main/docs/build-overview.md#output-directory-structure).
 - `VSC_clang` generates setup for Clang IntelliSense engine; for each *.cproject.yml file an additional file `.clang` is generated in the same directory that refers to the related `compile_commands.json` of the selected context.
 - `VSC_msft` generates setup for MSFT IntelliSense engine (tbd)

To build a project `cbuild <name>.csolution.yml -S` is called. Potentially the option `-C` for clean, or option `-R` for rebuild is used.

@mcgordonite would this fulfil the requirement for IDE setup?

**Comment by VGRSTM** [comment]
I've to shared my question mark here (sorry about ...). Is this code indexing concern part of OpenCMSIS project / initiative scope ?
What about end consumer doesn't targeting VSCode ?
Ok can be an option but even if not used by end consumer it increase code base so may have side effects.
Feel free but a bit ?? about myself.

**Comment by ReinhardKeil** [comment]
We could add over time other `--IDE <args>` to support other IDEs.  VS Code is just our initial target.

**Comment by mcgordonite** [comment]
The Csolution VS Code extension runs the `csolution convert` and `cbuild -t database` commands on start and each time the csolution input files change. There are three goals of this:
1. Generate a compile_commands.json file for each project in the selected `.buildType+targetType`.
2. Update RTE files for all contexts.
3. Update the cbuild-pack.yml file for all contexts.

Note that RTE files are updated for ALL contexts, but compile_commands.json files are only generated for the current context.

It then interacts with other extensions that provide intellisense in VS Code, based on the generated compile_commands.json files. Some of this is done via internal VS Code APIs, so I agree with @VGRSTM that this should be out of scope for CMSIS-Toolbox. Different IDEs will have different mechanisms for integrating knowledge from the build system into intellisense.

I think the question is whether we can streamline the commands mentioned above and still achieve the three goals. Some ideas:
* Would switching from `csolution convert` to `csolution update-rte` help?
* The IDE doesn't need to update RTE or cbuild-pack.yml files unless the user has changed the input files. This step could be omitted on first opening the IDE.
* Add a specific command, like the suggested `csolution setup`, that specifically runs the three required actions. Is this too specific to the VS Code use case?

> Do we need after this modification the "CMSIS_Project_Manager" anymore? Right now there are many different output windows created.

Possibly out of scope for this repository, but the two output channels can be merged together to simplify things.

**Comment by brondani** [comment]
Concerning the "Generator run information" I would suggest to (re-)use nodes that we already have:
https://github.com/Open-CMSIS-Pack/devtools/blob/f40190cfdc98ebc52f734b678deb0726a5a389b1/tools/projmgr/schemas/common.schema.json#L1059
For example:
```
   - component: Keil::Device:CubeMX@0.9.0
     generator:
       id: CubeMX
       path: <file>.cgen.yml
```

**Comment by jkrech** [comment]
For component's configuration files the cbuild.yml file currently lists:
```
 file: RTE/CMSIS/RTX_Config.h
 category: header
 attr: config
 version: 5.5.2
```

In addition it shall contain a PLM status whether the config file is:

 `<config>.<ext>.base@x.y.z` == config file version a.b.c in selected component version

```
  up-to-date :             x == a, y == b, z == c
  only patch differs:    x == a, y == b, z != c
  only minor differs:   x == a, y != b
  incompatible:           x != a
```

**Comment by brondani** [comment]
Talking to @mcgordonite and @JoeTheGuitarist we agree that having a `cbuild setup <name>.csolution.yml` command would improve the usability from the IDE perspective and would avoid multiple calls to different tools.

The operations performed by this command would be the following:
- for all contexts: if needed update RTE and `<name>.cbuild-pack.yml` files.
- for selected context: generate `<name>.cbuild.yml` and `compile_commands.json` files. To consider: `--context` option for processing one or multiple contexts at a time.

However we don't see much value in the proposed `--IDE` option that would be too specific for particular extensions.

**Comment by mcgordonite** [comment]
The summary looks good, apart from this line:
> create for VS Code the CLANG setup for Intellisense (when option --VSC_clang is used).

This flag sounds like the `--IDE` flag discussed in @brondani's [comment](https://github.com/Open-CMSIS-Pack/devtools/issues/1265#issuecomment-1912110128):
> However we don't see much value in the proposed --IDE option that would be too specific for particular extensions.

The compile_commands.json should be generated for all contexts in the context set, but this isn't specific to VS Code or clang.

**Comment by mcgordonite** [comment]
>The operations performed by this command would be the following:
> * for all contexts: if needed update RTE and <name>.cbuild-pack.yml files.
> * for selected context: generate <name>.cbuild.yml and compile_commands.json files. To consider: --context option for processing one or multiple contexts at a time.

After further discussion with @brondani & @jkrech, the RTE update does not need to occur for all contexts. It can be limited to the context set used to generated compile_commands.json and cbuild.yml files.

IDEs can offer a different mechanism to update RTE for all contexts using existing commands (e.g., `csolution update-rte`).

**Comment by jkrech** [comment]
To be solved by #1322
